run-python/runner.go

```diff
@@ -0,0 +1,187 @@
+		} else {
+			colon = strings.IndexByte(options.EntryPoint[equals+1:], ':')
+		}
+
+		if colon == -1 {
+			return nil, fmt.Errorf(
+				"expected entry-point of the form 'module[.submodule]:function', got: %v",
+				options.EntryPoint)
+		}
+
+		if equals == -1 {
+			options.EntryPoint = t.Name() + "=" + options.EntryPoint
+		}
+	}
+
+	root := filepath.Join(s.Root(), "python")
+
+	shellRunner, err := shell.NewRunner(s, t)
+	if err != nil {
+		return nil, err
+	}
+	sr, ok := shellRunner.(shell.Runner)
+	if !ok {
+		return nil, fmt.Errorf(
+			"unable to cast shell runner of type %T to shell.Runner",
+			shellRunner)
+	}
+	return runner{sr, root, pythonExecutable, options}, nil
+}
+
+func NewPythonRunner(s stoic.Stoic, t stoic.Tool) (tool.Runner, error) {
+	options := t.Config().Runner.Options
+	if _, ok := options["python"]; !ok {
+		options["python"] = "python"
+	}
+	return newRunner(s, t)
+}
+
+func NewPython2Runner(s stoic.Stoic, t stoic.Tool) (tool.Runner, error) {
+	t.Config().Runner.Options["python"] = "python2.7"
+	return newRunner(s, t)
+}
+
+func NewPython3Runner(s stoic.Stoic, t stoic.Tool) (tool.Runner, error) {
+	t.Config().Runner.Options["python"] = "python3"
+	return newRunner(s, t)
+}
+
+type PythonOptions struct {
+	Python           string `mapstructure:",omitempty"`
+	RequirementsFile string `mapstructure:"requirements,omitempty"`
+	ModulePath       string `mapstructure:"module-path,omitempty"`
+	EntryPoint       string `mapstructure:"entry-point,omitempty"`
+	Command          string `mapstructure:",omitempty"`
+}
+
+type runner struct {
+	ShellRunner shell.Runner
+
+	Root   string
+	Python string
+
+	PythonOptions
+}
+
+func (r runner) Setup(checkout tool.Checkout) error {
+	pe, err := newPythonEnvironment(r.Root, r.Python, r.ShellRunner.Stoic.Cache())
+	if err != nil {
+		return err
+	}
+
+	requirementsFile := filepath.Join(checkout.Path(), r.RequirementsFile)
+	ve, err := pe.NewVirtualEnvironment(requirementsFile)
+	if err != nil {
+		return err
+	}
+
+	setupCommand := exec.Command(ve.PythonExecutable(),
+		"-m", "pip", "check", "--quiet")
+
+	setupCommand.Stdout = os.Stdout
+	setupCommand.Stderr = os.Stderr
+	setupCommand.Env = ve.PythonEnviron()
+
+	err = setupCommand.Run()
+	if err != nil {
+		return err
+	}
+
+	r.ShellRunner.Options.SetupEnvironment["PIP_CACHE_DIR"] = pe.PipCache()
+	r.ShellRunner.Options.SetupEnvironment["PYTHONPATH"] = pe.SitePackages()
+	return r.ShellRunner.Setup(checkout)
+}
+
+func (r runner) Run(checkout tool.Checkout, name string, args []string) error {
+	pe, err := newPythonEnvironment(r.Root, r.Python, r.ShellRunner.Stoic.Cache())
+	if err != nil {
+		return err
+	}
+
+	requirementsFile := filepath.Join(checkout.Path(), r.RequirementsFile)
+	ve, err := pe.NewVirtualEnvironment(requirementsFile)
+	if err != nil {
+		return err
+	}
+
+	if r.EntryPoint != "" {
+		r.ShellRunner.Options.Parameters["EntryPoint"] = r.EntryPoint
+		r.ShellRunner.Options.Command = "{{.Python}} -c \"" +
```

Comment:
Replace with multiline string " -> `

Reply:
It's *not* multiline on purpose, as I didn't want to be passing multiline strings as command line arguments, but it'd probably be fine to do so. On the other hand, this tries to be short and sweet, so having unnecessary whitespace as part of the string would not be helpful.